### ISSUE TITLE
Questions for the class prediction and label annotation

### ISSUE THREAD

**Issue by flow233-ai** [coperception/coperception#23]
I am confused about the question about the class prediction and label annotation:
```python
per_entry_cross_ent = _softmax_cross_entropy_with_logits(
            labels=target_tensor, logits=prediction_tensor
        )

        # convert [N, num_anchors] to [N, num_anchors, num_classes]
        per_entry_cross_ent = per_entry_cross_ent.unsqueeze(-1) * target_tensor
        prediction_probabilities = F.softmax(prediction_tensor, dim=-1)
        p_t = (target_tensor * prediction_probabilities) + (
            (1 - target_tensor) * (1 - prediction_probabilities)
        )
```
First, as shown above，the class `SoftmaxFocalClassificationLoss` in `loss.py` contain these code lines. The `_softmax_cross_entropy_with_logits` function uses `prediction_tensor` and `target_tensor` as input. But from my view, the prediction_tensor should firstly be compressed to [0,1] using `F.softmax` and then be used as the input. So I think this function should use `prediction_probabilities` and `target_tensor` as input.
```python
data = {
            "bev_seq": padded_voxel_points.to(device),
            "labels": label_one_hot.to(device),
            "reg_targets": reg_target.to(device),
            "anchors": anchors_map.to(device),
            "vis_maps": vis_maps.to(device),
            "reg_loss_mask": reg_loss_mask.to(device).type(dtype=torch.bool),
            "target_agent_ids": target_agent_ids.to(device),
            "num_agent": num_all_agents.to(device),
            "trans_matrices": trans_matrices.to(device),
        }
```
Second, I find that the shape of variable `label_one_hot` is [6,256,256,6,2]. The dim0=6 is the num of agent. The dim1,dim2=256 is the num of voxels. But I am confused about what the last two dimension 6 and 2 stand for. Maybe now I'm not very familiar with vehicle perception, but I hope you can help answer it.

Loonking forward to your reply. Thanks a lot!

**Comment by dekunma** [comment]
For your first question, I think @RoboticsYimingLi could probably give a more precise answer.  

For your second question, as the variable name suggests, the labels are one-hot encoded, so the last two dimensions are one-hot labels for each voxel for each agent. 
For example, `label_one_hot[0][0][0]` could be 
```
[[1, 0],
[1, 0],
[1, 0],
[1, 0],
[1, 0],
[1, 0]]
```
The first dimension 6 means there are 6 agents in the dataset, and the second dimension 2 means there are 2 categories, as specified in 
https://github.com/coperception/coperception/blob/7e99052b9d9686002482e28cbf5179cb5b8cddfe/coperception/configs/Config.py#L172-L173
So, it means that the label for agent `0` at its voxelized birds-eye-view point cloud position `[0, 0]` is `0`.

**Comment by flow233-ai** [comment]
Thanks for your answer.
As for the first question, I will contact [https://github.com/RoboticsYimingLi](url) for more precise answer.
But for the second question, I am still confused. As I mentioned above, the shape of variable `label_one_hot` is `[6,256,256,6,2]`.  If the foruth dimenstion means there are 6 agents in the dataset, what does the first dimenstion mean? You mean they are both standing for 6 agents in the dataset?

**Comment by dekunma** [comment]
The first dimension actually means the number of agents involved in the current training process. If you set `rsu` to `0` to exclude RSU, then it will be `5`.  

---
> the six elements are always the same.  

Yes, they should always be the same because the ground truth for a voxel should be the same for all agents.

**Comment by flow233-ai** [comment]
I got it. Thanks again for your meticulous answer！

**Comment by simbaforrest** [comment]
@flow233-ai Is your issue addressed? If so, please close it. Thank you.

**Comment by dekunma** [comment]
@flow233-ai Since this issue seems to be resolved, I'm closing it now. Feel free to reopen it if you have any follow-up questions.
